flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,340 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
+               ? liveIn.getDefiningOp()->getResult(0)
+               : liveIn;
+  };
+
+  // Re-package both lists of live-ins into sets so that we can use set equality
+  // to compare the values used in the outerloop vs. the nestd one.
+
+  for (auto liveIn : nestedLiveIns)
+    nestedLiveInsSet.insert(getUnifiedLiveInView(liveIn));
+
+  mlir::Value outerLoopIV;
+  for (auto liveIn : outerLoopLiveIns) {
+    outerLiveInsSet.insert(getUnifiedLiveInView(liveIn));
+
+    // Keep track of the IV of the outerloop. See `isPerfectlyNested` for more
+    // info on the reason.
+    if (outerLoopIV == nullptr)
+      outerLoopIV = getUnifiedLiveInView(liveIn);
+  }
+
+  // For the 2 loops to be perfectly nested, either:
+  // * both would have exactly the same set of live-in values or,
+  // * the outer loop would have exactly 1 extra live-in value: the outer
+  //   loop's induction variable; this happens when the outer loop's IV is
+  //   *not* referenced in the nested loop.
+  bool isPerfectlyNested = [&]() {
+    if (outerLiveInsSet == nestedLiveInsSet)
+      return true;
+
+    if ((outerLiveInsSet.size() == nestedLiveIns.size() + 1) &&
+        !nestedLiveInsSet.contains(outerLoopIV))
+      return true;
+
+    return false;
+  }();
+
+  if (!isPerfectlyNested)
+    return mlir::failure();
+
+  return collectLoopNest(nestedUnorderedLoop, nestedLiveIns, loopNest);
+}
+
+/// Prepares the `fir.do_loop` nest to be easily mapped to OpenMP. In
+/// particular, this function would take this input IR:
+/// ```
+/// fir.do_loop %i_iv = %i_lb to %i_ub step %i_step unordered {
+///   fir.store %i_iv to %i#1 : !fir.ref<i32>
+///   %j_lb = arith.constant 1 : i32
+///   %j_ub = arith.constant 10 : i32
+///   %j_step = arith.constant 1 : index
+///
+///   fir.do_loop %j_iv = %j_lb to %j_ub step %j_step unordered {
+///     fir.store %j_iv to %j#1 : !fir.ref<i32>
+///     ...
+///   }
+/// }
+/// ```
+///
+/// into the following form (using generic op form since the result is
+/// technically an invalid `fir.do_loop` op:
+///
+/// ```
+/// "fir.do_loop"(%i_lb, %i_ub, %i_step) <{unordered}> ({
+/// ^bb0(%i_iv: index):
+///   %j_lb = "arith.constant"() <{value = 1 : i32}> : () -> i32
+///   %j_ub = "arith.constant"() <{value = 10 : i32}> : () -> i32
+///   %j_step = "arith.constant"() <{value = 1 : index}> : () -> index
+///
+///   "fir.do_loop"(%j_lb, %j_ub, %j_step) <{unordered}> ({
+///   ^bb0(%new_i_iv: index, %new_j_iv: index):
+///     "fir.store"(%new_i_iv, %i#1) : (i32, !fir.ref<i32>) -> ()
+///     "fir.store"(%new_j_iv, %j#1) : (i32, !fir.ref<i32>) -> ()
+///     ...
+///   })
+/// ```
+///
+/// What happened to the loop nest is the following:
+///
+/// * the innermost loop's entry block was updated from having one operand to
+///   having `n` operands where `n` is the number of loops in the nest,
+///
+/// * the outer loop(s)' ops that update the IVs were sank inside the innermost
+///   loop (see the `"fir.store"(%new_i_iv, %i#1)` op above),
```

Comment:
What happens with other in-between code?

Reply:
The other in-between code here would be the ops needed to setup the LB, UB, and step of loops below the outermost one. These are cloned by `genLoopNestClauseOps(...)`.

Since we only support perfectly nested loops (see `collectLoopNest -> isPerfectlyNested`), there will be no other in-between code.

Reply:
👍
Could you mention this assumes there is no other in-between code?